aws/rust-runtime/aws-auth/src/refresh.rs

```diff
@@ -0,0 +1,294 @@
+        let watchdog_signal = cancel_signal.clone();
+        thread::Builder::new()
+            .stack_size(WATCHDOG_STACK_SIZE_BYTES)
+            .name("RefreshingCredentialProvider-watchdog".into())
+            .spawn(move || {
+                Self::watchdog_thread(watchdog_signal, inner);
+            })
+            .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))?;
+
+        Ok(CancelableRefreshLoop {
+            signal: Some(cancel_signal),
+        })
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    fn watchdog_thread(signal: RwCell<RefreshLoopState>, inner: Inner) {
+        while let RefreshLoopState::Running = signal.get() {
+            // Spawn an executor thread to run user-provided async credential refresher.
+            // This needs to have its own thread so that a panic won't end the refresh loop.
+            let executor_signal = signal.clone();
+            let executor_inner = inner.clone();
+            let result = thread::Builder::new()
+                .name("RefreshingCredentialProvider-executor".into())
+                .spawn(move || {
+                    RefreshingCredentialProvider::executor_thread(executor_signal, executor_inner);
+                })
+                .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))
+                .and_then(|joiner| joiner.join().map_err(|_| Error::ExecutorThreadFailed));
+            if let Err(err) = result {
+                error!("RefreshingCredentialProvider executor failed: {}", err);
+                thread::sleep(EXECUTOR_PANIC_BACKOFF);
+            }
+        }
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    fn executor_thread(signal: RwCell<RefreshLoopState>, inner: Inner) {
+        let runtime = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("failed to start tokio runtime");
+        runtime.block_on(async move {
+            // TODO: Take a name argument from users to indicate which credentials are being refreshed?
+            let span = trace_span!("credential_refresh_loop");
+            let _enter = span.enter();
+
+            info!("Starting credential refresh loop");
+            Self::refresh_loop(signal, inner)
+                .await
+                .expect("refresh loop failed");
+        });
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    async fn refresh_loop(signal: RwCell<RefreshLoopState>, inner: Inner) -> Result<(), Error> {
+        while let RefreshLoopState::Running = signal.get() {
+            while let Some(sleep_duration) = inner.next_refresh_sleep_time(SystemTime::now()) {
+                tokio::time::sleep(sleep_duration).await;
+            }
+
+            info!("Refreshing credentials");
+            let timeout_future = tokio::time::sleep(REFRESH_TIMEOUT);
+            let refresh_future = inner.refresh();
+            tokio::pin!(timeout_future, refresh_future);
+            loop {
```

Comment:
What is this `loop` for? It looks like all `select!` branches break, so there are no flows that loop?